### ISSUE TITLE
Errors when ReFrame root directory is not first path in `sys.path`

### ISSUE THREAD

**Issue by mahendrapaipuri** [eth-cscs/reframe#2135]
Hello,

The context here is when I use ReFrame CI integration capabilities, I am getting errors in the dependent jobs. When we use `--restore-session` flag for dependent job, ReFrame is detecting modules from filenames [here](https://github.com/eth-cscs/reframe/blob/bb779ce73a4f081aef1b6e0c20c127061bec24d1/reframe/utility/__init__.py#L84). 

I am not sure of the rabbit hole details of why it is being done, but the issue I have noticed is that on that stated line, we implicitly assume `sys.path[0]` if ReFrame root directory. If it is not, ReFrame will have problems finding the modules. For example, in this [failed CI pipeline](https://gitlab.com/ska-telescope/sdp/ska-sdp-benchmark-tests/-/jobs/1512352442#L87), I printed the `sys.path` and we can see that ReFrame root directory is not the first path.

This is the directory structure of my repository:
```
.
├── apps
├── config
├── docs
├── helloworld
├── modules
├── output
├── perflogs
├── reframe
├── spack
└── topologies
```

And I usually launch tests from root of the repo using:
`reframe/bin/reframe -C <config> -c <test> -r`

**Comment by mahendrapaipuri** [comment]
Follow up on this issue. I could fix by [pipelines](https://gitlab.com/ska-telescope/sdp/ska-sdp-benchmark-tests/-/pipelines/355090332) after bit of a hack.

So, I am iterating over all the paths in `sys.path` and looking for ReFrame root and moving it to the front in my test script.
```
# ReFrame root needs to be the first path for ReFrame to work properly
# First make a copy of sys.path
sys_path_copy = sys.path.copy()
for path in sys_path_copy:
    root = os.path.basename(os.path.normpath(path))
    # if reframe root is found, remove it from list and add it to front
    if root == 'reframe':
        sys.path.remove(path)
        sys.path.insert(0, path)
        break
```
This solved all my CI issues. Seems like having ReFrame root as the first path in `sys.path` is critical for CI jobs to succeed. It might be worth doing something like this (my hack is dirty. I m pretty sure it can be done more elegantly) in `import_module_from_file` function in `reframe.utility`.

**Comment by rsarm** [comment]
Hi @mahendrapaipuri. The CI pipeline linked above didn't fail, right? Or I'm looking to the wrong place? If that's the case, could you please share again one of the ones that failed?

In the reframe executable [the reframe dir and the external are prepended to the `sys.path`](https://github.com/eth-cscs/reframe/blob/792939c69630e9e936ccafa275487512a9b5bbad/bin/reframe#L12-L16). That's why later the reframe dir is in `sys.path[0]`.

**Comment by mahendrapaipuri** [comment]
Hello @rsarm, thanks for coming back to the issue. Actually, the pipeline failed. Let me try to put things more constructively:

- Thanks for pointing me to ReFrame executable and yes, I agree that ReFrame should be on the front in `sys.path`. However, I printed the `sys.path` on one of the failed pipelines -> https://gitlab.com/ska-telescope/sdp/ska-sdp-benchmark-tests/-/jobs/1512352442#L91 and we can see here that ReFrame is not the first one in `sys.path`. This job is the one that generates child pipeline and it always runs successfully. It is the next job, ie the one that executes the child pipeline, that fails. Please have a look at it -> https://gitlab.com/ska-telescope/sdp/ska-sdp-benchmark-tests/-/jobs/1512502926 .
- I am not quite sure why `/builds/ska-telescope/sdp/ska-sdp-benchmark-tests` path popped up in there. I have noticed that it happened only in the CI jobs. Maybe it is something GitLab CI specific? I never faced this issue when running the tests on HPC platforms.
- In any case, we can notice from this failed job that we can indeed mess up `sys.path` which can fail ReFrame. That is why I suggested to check the `sys.path` in `import_module_from_file` function to make it more resilient.

Is it more clear now? If not, we can have a quick chat about the issue. Cheers

**Comment by mahendrapaipuri** [comment]
Just to give more info on the issue. I guess the issue comes from here -> https://gitlab.com/ska-telescope/sdp/ska-sdp-benchmark-tests/-/blob/93048cf173e1c5c0b3cafbbdc11c309297c1d5ad/apps/level1/imaging-iotest/reframe_iotest.py#L63 . I do not remember why, back then I was appending the current working directory path to `sys.path` as pointed out in the line. That is why `/builds/ska-telescope/sdp/ska-sdp-benchmark-tests` is on the front in `sys.path`. 

And why it failed only in CI? Well, I guess `import_module_from_file` function is called only when invoking ReFrame with `--report-file` and/or `--restore-session` flags, if I remember well. In the child pipeline CI jobs, ReFrame is invoked with those flags -> https://gitlab.com/ska-telescope/sdp/ska-sdp-benchmark-tests/-/jobs/1512502926#L46 . I never used these flags when running tests on HPC systems and that is why I did not face this issue when when ReFrame is not on front in `sys.path`.

I hope that helps!

**Comment by mahendrapaipuri** [comment]
Here is a MWE to reproduce the bug. 

A sample ReFrame test with dependencies:

```
import os
import reframe as rfm
import reframe.utility.sanity as sn
import reframe.utility.udeps as udeps

@rfm.simple_test
class ReadFileTest(rfm.RunOnlyRegressionTest):
    descr = 'Test to read the created file'
    valid_systems = ['*']
    valid_prog_environs = ['*']

    @run_after('init')
    def inject_dependencies(self):
        self.depends_on('FileCreateTest', udeps.fully)

    @require_deps
    def set_sourcedir(self, FileCreateTest):
        self.sourcesdir = os.path.join(
            FileCreateTest().stagedir
        )

    @run_before('run')
    def read_file(self):
        self.executable = 'cat'
        self.executable_opts = [os.path.join(self.sourcesdir, 'test.txt')]

    @run_before('sanity')
    def set_sanity_patterns(self):
        self.sanity_patterns = sn.assert_not_found('error', self.stderr)


@rfm.simple_test
class FileCreateTest(rfm.RunOnlyRegressionTest):
    descr = 'Test to create a sample text file'
    valid_systems = ['*']
    valid_prog_environs = ['*']
    executable = 'echo "This is sample text file" > test.txt'
    
    @run_before('sanity')
    def set_sanity_patterns(self):
        self.sanity_patterns = sn.assert_not_found('error', self.stderr)
```
A generic ReFrame config file can run this test on any machine:

```
import os
import sys

# Add root dir to PYTHONPATH
try:
    root_dir = os.path.dirname(os.path.realpath(__file__))
    sys.path.index(root_dir)
except ValueError:
    sys.path.insert(0, root_dir)


site_configuration = {
    'systems': [
        {
            'name': 'generic',
            'descr': 'Generic system. When no system is matched, it falls back to generic system',
            'hostnames': ['.*'],
            'partitions': [
                {
                    'name': 'default',
                    'scheduler': 'local',
                    'launcher': 'local',
                    'environs': ['builtin']
                }
            ]
        },  # end generic
    ],
    'environments': [
        {
            'name': 'builtin',
            'cc': 'gcc',
            'cxx': 'g++',
            'ftn': 'gfortran',
            'target_systems': [
                'generic',
            ]
        },
    ],
}
```
**Notice** that we are inserting the current directory path to the **beginning** of the `sys.path` in the config file. This is **important** to reproduce the bug.

Normally, we can run this test using
`reframe/bin/reframe -C reframe_config.py -c <test_file>.py -r` command. However, when we want to run this in CI by invoking `--generate-ci` flag, ReFrame resolves the dependencies by creating report file for each test and using the created report file to restore session in the dependent test as follows:

```
# Run first test
reframe/bin/reframe -C reframe_config.py -c test.py --report-file=FileCreateTest-report.json -n FileCreateTest -r 

# Run dependent second test
reframe/bin/reframe -C reframe_config.py -c test.py --report-file=ReadFileTest-report.json --restore-session=FileCreateTest-report.json -n ReadFileTest -r  
```
If we run it this way, we will see that first test runs successfully but the second test fails with following error

```
reframe/bin/reframe: run session stopped: module not found error: No module named 'reframe.reframe'
reframe/bin/reframe: Traceback (most recent call last):
  File "/home/mahendra/Dropbox/IR_INRIA/Scripts/ReFrame/reframe/reframe/frontend/cli.py", line 878, in main
    testgraph, restored_cases = report.restore_dangling(testgraph)
  File "/home/mahendra/Dropbox/IR_INRIA/Scripts/ReFrame/reframe/reframe/frontend/runreport.py", line 96, in restore_dangling
    self._do_restore(d)
  File "/home/mahendra/Dropbox/IR_INRIA/Scripts/ReFrame/reframe/reframe/frontend/runreport.py", line 111, in _do_restore
    testcase._check = jsonext.load(fp)
  File "/home/mahendra/Dropbox/IR_INRIA/Scripts/ReFrame/reframe/reframe/utility/jsonext.py", line 134, in load
    return json.load(fp, **kwargs)
  File "/home/mahendra/miniconda3/envs/ska-sdp-benchmark-tests/lib/python3.8/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/home/mahendra/miniconda3/envs/ska-sdp-benchmark-tests/lib/python3.8/json/__init__.py", line 370, in loads
    return cls(**kw).decode(s)
  File "/home/mahendra/miniconda3/envs/ska-sdp-benchmark-tests/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/home/mahendra/miniconda3/envs/ska-sdp-benchmark-tests/lib/python3.8/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "/home/mahendra/Dropbox/IR_INRIA/Scripts/ReFrame/reframe/reframe/utility/jsonext.py", line 98, in _object_hook
    mod = util.import_module_from_file(filename)
  File "/home/mahendra/Dropbox/IR_INRIA/Scripts/ReFrame/reframe/reframe/utility/__init__.py", line 101, in import_module_from_file
    return importlib.import_module(module_name)
  File "/home/mahendra/miniconda3/envs/ska-sdp-benchmark-tests/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'reframe.reframe'

```
This is due to the fact that we messed up the `sys.path` and ReFrame starts to look for its module in non-existent places. **We can only reproduce** this bug when we use the report file from preceding test to restore session in the dependent test. 

For instance, if we change the path manipulation in config file as `sys.path.insert(1, root_dir)`, *i.e.*, inserting the path in second position, the tests run normally without any errors.

**Comment by vkarak** [comment]
The reproducer for my case gives the following:

```
/Users/karakasv/Repositories/reframe/bin/reframe: run session stopped: config error: 'local' is already registered as a scheduler
```

**Comment by vkarak** [comment]
But it works with the fix in #2201.
